sql/analyzer/optimize_joins.go

```diff
@@ -0,0 +1,364 @@
+			leftTableExpr, err := fixFieldIndexes(node.Left.Schema(), cond.Left())
+			if err != nil {
+				return nil, nil, nil, nil, err
+			}
+			return node.Left, node.Right, leftTableExpr, idx, nil
+		}
+		if joinType != plan.JoinTypeLeft && indexMatches(idx, unifyExpression(aliases, cond.Left())) {
+			rightTableExpr, err := fixFieldIndexes(node.Right.Schema(), cond.Right())
+			if err != nil {
+				return nil, nil, nil, nil, err
+			}
+			return node.Right, node.Left, rightTableExpr, idx, nil
+		}
+	}
+
+	return nil, nil, nil, nil, errors.New("couldn't determine suitable indexes to use for tables")
+}
+
+// indexMatches returns whether the given index matches the given expression using the expression's string
+// representation. Compare to logic in IndexRegistry.IndexByExpression
+func indexMatches(index sql.Index, expr sql.Expression) bool {
+	if len(index.Expressions()) != 1 {
+		return false
+	}
+
+	indexExprStr := index.Expressions()[0]
+	return indexExprStr == expr.String()
+}
+
+// Returns the underlying table name for the node given
+func findTableName(node sql.Node) string {
+	var tableName string
+	plan.Inspect(node, func(node sql.Node) bool {
+		switch node := node.(type) {
+		case *plan.ResolvedTable:
+			// TODO: this is over specific, we only need one side of the join to be indexable
+			if it, ok := node.Table.(sql.IndexableTable); ok {
+				tableName = it.Name()
+				return false
+			}
+		}
+		return true
+	})
+
+	return tableName
+}
+
+// Returns the table and field names from the expression given
+func getTableNameFromExpression(expr sql.Expression) (tableName string, fieldName string) {
+	switch expr := expr.(type) {
+	case *expression.GetField:
+		tableName = expr.Table()
+		fieldName = expr.Name()
+	}
+
+	return tableName, fieldName
+}
+
+// index munging
+
+// Assign indexes to the join conditions and returns the sql.Indexes assigned, as well as returning any aliases used by
+// join conditions
+func findJoinIndexes(ctx *sql.Context, a *Analyzer, node sql.Node) ([]sql.Index,  Aliases, error) {
+	a.Log("finding indexes, node of type: %T", node)
+
+	indexSpan, _ := ctx.Span("find_join_indexes")
+	defer indexSpan.Finish()
+
+	var indexes []sql.Index
+	// release all unused indexes
+	defer func() {
+		if indexes == nil {
+			return
+		}
+
+		for _, index := range indexes {
+			a.Catalog.ReleaseIndex(index)
+		}
+	}()
+
+	aliases := make(Aliases)
+	var err error
+
+	fn := func(ex sql.Expression) bool {
+		if alias, ok := ex.(*expression.Alias); ok {
+			if _, ok := aliases[alias.Name()]; !ok {
+				aliases[alias.Name()] = alias.Child
+			}
+		}
+		return true
+	}
+
+	plan.Inspect(node, func(node sql.Node) bool {
+		switch node := node.(type) {
+		case *plan.InnerJoin, *plan.LeftJoin, *plan.RightJoin:
+
+			var cond sql.Expression
+			switch node := node.(type) {
+			case *plan.InnerJoin:
+				cond = node.Cond
+			case *plan.LeftJoin:
+				cond = node.Cond
+			case *plan.RightJoin:
+				cond = node.Cond
+			}
+
+			fn(cond)
```

Comment:
Shouldn't this visit the expression tree or something? I don't understand how this sees an `Alias` almost ever?

Reply:
Yes, you are right. This should probably fail on queries that use aliases. I'm going to leave this alone for now and try to write some tests that fail.